tests/arrays.rs

```diff
@@ -1,212 +1,137 @@
-struct AdditionalSettings {
-    inner: u8,
+#[derive(Debug, Copy, Clone, Default, Miniconf, Deserialize, Serialize)]
+struct Inner {
+    c: u8,
 }
 
-#[derive(Debug, Default, Miniconf, Deserialize)]
+#[derive(Debug, Default, Miniconf)]
 struct Settings {
-    data: u32,
+    a: [u8; 2],
     #[miniconf(defer)]
-    more: AdditionalSettings,
+    d: [u8; 2],
+    #[miniconf(defer)]
+    dm: [Inner; 2],
+    #[miniconf(defer)]
+    am: miniconf::Array<Inner, 2>,
+    #[miniconf(defer)]
+    aam: miniconf::Array<miniconf::Array<Inner, 2>, 2>,
 }
 
 #[test]
-fn simple_array() {
-    #[derive(Miniconf, Default)]
-    struct S {
-        #[miniconf(defer)]
-        a: [u8; 3],
-    }
-
-    let mut s = S::default();
-
-    // Updating a single field should succeed.
-    s.set("/a/0", "99".as_bytes()).unwrap();
-    assert_eq!(99, s.a[0]);
-
-    // Updating entire array atomically is not supported.
-    assert!(s.set("/a", "[1,2,3]".as_bytes()).is_err());
-
-    // Invalid index should generate an error.
-    assert!(s.set("/a/100", "99".as_bytes()).is_err());
+fn atomic() {
+    let mut s = Settings::default();
+    s.set_json("/a", b"[1,2]").unwrap();
+    assert_eq!(s.a, [1, 2]);
 }
 
 #[test]
-fn nonexistent_field() {
-    #[derive(Miniconf, Default)]
-    struct S {
-        #[miniconf(defer)]
-        a: [u8; 3],
-    }
-
-    let mut s = S::default();
-
-    assert!(s.set("/a/1/b", "7".as_bytes()).is_err());
+fn defer() {
+    let mut s = Settings::default();
+    s.set_json("/d/1", b"99").unwrap();
+    assert_eq!(s.d[1], 99);
 }
 
 #[test]
-fn simple_array_indexing() {
-    #[derive(Miniconf, Default)]
-    struct S {
-        #[miniconf(defer)]
-        a: [u8; 3],
-    }
-
-    let mut s = S::default();
-
-    s.set("/a/1", "7".as_bytes()).unwrap();
-
-    assert_eq!([0, 7, 0], s.a);
-
-    // Ensure that setting an out-of-bounds index generates an error.
-    assert!(matches!(
-        s.set("/a/3", "7".as_bytes()).unwrap_err(),
-        Error::BadIndex
-    ));
-
-    // Test metadata
-    let metadata = S::metadata(1);
-    assert_eq!(metadata.max_depth, 2);
-    assert_eq!(metadata.max_length, "/a/2".len());
-    assert_eq!(metadata.count, 3);
+fn defer_miniconf() {
+    let mut s = Settings::default();
+    s.set_json("/am/0/c", b"1").unwrap();
+    assert_eq!(s.am[0].c, 1);
+    s.set_json("/aam/0/0/c", b"3").unwrap();
+    assert_eq!(s.aam[0][0].c, 3);
 }
 
 #[test]
-fn array_iter() {
-    #[derive(Miniconf, Default, Clone, Copy, Debug, PartialEq)]
-    struct Inner {
-        b: u8,
-    }
-
-    #[derive(Miniconf, Default)]
-    struct S {
-        #[miniconf(defer)]
-        a: miniconf::Array<miniconf::Array<Inner, 2>, 2>,
-    }
-
-    let mut s = S::default();
-
-    for _i in s.a.into_iter().flatten() {}
-
-    for _i in s.a.iter().flatten() {}
-
-    for _i in s.a.iter_mut().flatten() {}
+fn too_short() {
+    let mut s = Settings::default();
+    assert_eq!(s.set_json("/d", b"[1,2,3]"), Err(Error::Internal(1)));
```

Comment:
```suggestion
    assert_eq!(s.set_json("/d", b"[1,2]"), Err(Error::Internal(1)));
```

Reply:
This also tests precedence of `Internal` over `Inner`

Reply:
Ah, gotcha, I didn't realize that when reviewing. Can you add a note? It looks like a bug otherwise.

Reply:
Made it explicit.